### ISSUE TITLE
[feature request] LatexPrinter: comma as decimal separator

### ISSUE THREAD

**Issue by gschintgen** [sympy/sympy#16565]
Hi,

Normally I don't mind the use of a period as decimal separator (not at all in fact), but when using Sympy in LaTeX documents via PythonTeX, this is becoming quite problematic, since the generated documents have to be in French and decimal numbers are usually written with a comma as decimal separator.

Unfortunately LatexPrinter does not seem to have some kind of "decimal_separator" option, hence my request: it'd be very much appreciated if I could switch the decimal separator for latex'ed output.

(If this is implemented, it might be advisable to simultaneously replace the list item separator from comma to semicolon if decimal_separator='comma' is active.)

**Comment by oscarbenjamin** [comment]
That seems like a reasonable feature request.

Can you give an example of the output that you currently see and what you would like it to be? Also can you show what you mean about the list item separator?

**Comment by Sc0rpi0n101** [comment]
I think he means lists will be like:
``A = [1; 2; 3]``
instead of 
``A = [1, 2, 3]``

when decimal_separatior = comma

**Comment by gschintgen** [comment]
Well, currently, I'm mostly printing single float values, i.e. numerical approximations given by .n() or .round(). These show up as 0.134 (wanted: 0,134) or '4.76190476190476 \\cdot 10^{-7}' where I'd like the same output but starting with 4,7619...

But obviously such a setting should also apply to floating point coefficients in a larger symbolic expression.

Do note that in (La)TeX it's necessary to put the comma between braces, otherwise a small but superfluous space is inserted, as if the integer and fractional parts where two unrelated elements of a set or items in a list. Ideally LatexPrinter would either give '3.14' or '3{,}14' as latex-output depending on the value given to the requested decimal_separator option.

As for the list item separator, in regions where a comma is used as a decimal separator it's often customary to use a semicolon for lists of numbers in order to avoid confusion: It's bad style to use a comma both for separating list entries and as a decimal separator, at least if the list items are in decimal notation.
For example, the list [1, 2.718, 3.14] is currently latex()ed as:
'\\left [ 1, \\quad 2.718, \\quad 3.14\\right ]'
What I'm suggesting for lists (at least with numeric content) is:
'\\left [ 1; \\quad 2{,}718; \\quad 3{,}14\\right ]'
But I do realise that with the rather large \quad spaces it's not nearly as bad as it would be with a smaller regular space character.

The main issue I'm having is the hard-coded period character as decimal separator.

**Comment by asmeurer** [comment]
This should be easy to fix.

**Comment by gschintgen** [comment]
Thanks for the quick responses.
(The same reasoning would of course also apply to sets and tuples and not only to lists. That part of the request may be debatable, since it may be considered to interfere with Python's look and feel, but personally I consider the latex'ed output to be math rather than some python-syntax.)

**Comment by Sc0rpi0n101** [comment]
Can't ``expr.replace(",", ".")`` work?

you replace the ``.`` with ``,`` in the expression then print it

**Comment by oscarbenjamin** [comment]
@gschintgen can you be very clear what you men showing examples of input and output?

The format of an interactive session in IPython helps make this clear e.g. right now we have:
```julia
In [23]: print(latex([1.2, 3.4]))                                                                                                              
\left[ 1.2, \  3.4\right]
```
I guess that you would like to call this like `latex(..., decimal_separator='comma')` so can you show exactly how it *would* look using the hypothetical changed code in IPython? The more examples the better.

**Comment by gschintgen** [comment]
Below is the CURRENT output I get on the command line with the current git version.
(Previous output was from an older version running in a docker container with sagemath. I can't mess with it right now.)

```
Python 3.6.7 (default, Oct 22 2018, 11:32:17)
[GCC 8.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from sympy import *
>>> latex([1, 2.3, 4.5])
'\\left[ 1, \\  2.3, \\  4.5\\right]'
>>> latex(FiniteSet(1, 2.3, 4.5))
'\\left\\{1, 2.3, 4.5\\right\\}'
>>> latex((1, 2.3, 4.5))
'\\left( 1, \\  2.3, \\  4.5\\right)'
>>> latex(0.987)
'0.987'
>>> latex(S(0.987))
'0.987'
>>> latex(.3)
'0.3'
>>> latex(S(.3))
'0.3'
>>> latex(5.8*10**(-7))
'5.8e-07'
>>> latex(S(5.7)*10**(-7))
'5.7 \\cdot 10^{-7}'
>>> latex(S(5.7*10**(-7)))
'5.7 \\cdot 10^{-7}'
>>> var('x')
x
>>> latex(1.2*x+3.4)
'1.2 x + 3.4'

```
And here is a FICTITIOUS output of the same session:
```
>>> latex([1, 2.3, 4.5], decimal_separator='comma')
'\\left[ 1; \\  2{,}3; \\  4{,}5\\right]'
>>> latex(FiniteSet(1, 2.3, 4.5), decimal_separator='comma')
'\\left\\{1; 2{,}3; 4{,}5\\right\\}'
>>> latex((1, 2.3, 4.5))
'\\left( 1; \\  2{,}3, \\  4{,}5\\right)'
>>> latex(0.987, decimal_separator='comma')
'0{,}987'
>>> latex(S(0.987), decimal_separator='comma')
'0{,}987'
>>> latex(.3, decimal_separator='comma')
'0{,}3'
>>> latex(S(.3), decimal_separator='comma')
'0{,}3'
>>> latex(5.8*10**(-7), decimal_separator='comma')
'5{,}8e-07'          <------------- ?????? (not sure about this)
>>> latex(S(5.7)*10**(-7), decimal_separator='comma')
'5{,}7 \\cdot 10^{-7}'
>>> latex(S(5.7*10**(-7)), decimal_separator='comma')
'5{,}7 \\cdot 10^{-7}'
>>> var('x')
x
>>> latex(1.2*x+3.4, decimal_separator='comma')
'1{,}2 x + 3{,}4'

```
The `{,}` syntax may seem strange, but is the customary solution for plain (La)TeX. There is of course a LaTeX package (icomma) that takes care of this minor spacing issue automatically (i.e. a plain comma inbetween digits will no longer insert a small space), but it does not convert periods to commas.

As an aside, I noticed the following:
```
>>> latex(5.6*10**(-7))
'5.599999999999999e-07'
>>> latex(S(5.6)*10**(-7))
'5.6 \\cdot 10^{-7}'
>>> latex(S(5.6*10**(-7)))
'5.6 \\cdot 10^{-7}'
```
I'm wondering if the first printout is the intended behaviour. I do understand that the internal float value (with its Python type instead of a Sympy type) may be unable to represent exactly 5.6, but I (naïvely) expected Sympy's latex function to always sympify its input.

**Comment by gschintgen** [comment]
> 
> 
> Can't `expr.replace(".", ",")` work?
> 
> you replace the `.` with `,` in the expression then print it

This is indeed a possible workaround. The reason I'm suggesting a new feature is that it would be in line with other typographic options supported by LatexPrinter. What's more, PythonTeX has convenience macros like `\sympy{.....}` that automatically take care of calling latex() and doing the necessary handling. Furthermore PythonTeX already has an interface that lets the user set the appropriate LatexPrinter options. So it'd be the most natural solution to have a specific option for this.

**Comment by oscarbenjamin** [comment]
Thanks @gschintgen for the detailed examples. That will be very helpful for someone who tries to implement this feature.

The other reason I asked for you to do that because it would highlight some problematic cases and you've identified the one that I expected which is numbers of the form 1.2e10. In the case
```julia
In [48]: latex(5.8e-7)                                                                                                                         
Out[48]: '5.8e-07'
```
What is it that you are not sure about? Is it common to use a comma with numbers in this form? Or is it that you think `latex()` should show this with `\cdot 10^{-7}`?

For the other case you mention note that
```julia
In [51]: 5.6*10**-7 == 5.6e-7                                                                                                                  
Out[51]: False
```
since the way that you create the number is leading to a rounding error. This is compounded by a bug in SymPy's `Float.__str__`:
```julia
In [53]: x1 = S(5.6*10**-7)                                                                                                                    

In [54]: x2 = S('5.6e-7')                                                                                                                      

In [55]: x1                                                                                                                                    
Out[55]: 5.60000000000000e-7

In [56]: x2                                                                                                                                    
Out[56]: 5.60000000000000e-7

In [57]: str(x1) == str(x2)                                                                                                                    
Out[57]: True

In [58]: x1 == x2                                                                                                                              
Out[58]: False

In [61]: srepr(x1)                                                                                                                             
Out[61]: "Float('5.5999999999999993e-7', precision=53)"

In [62]: srepr(x2)                                                                                                                             
Out[62]: "Float('5.6000000000000004e-7', precision=53)"
```
For Python floats we have the property that `eval(str(x)) == x` for all floats x (except nan). SymPy's `Float` often breaks this property. I think there already is an open issue somewhere about this.

**Comment by gschintgen** [comment]
> ```julia
> In [48]: latex(5.8e-7)                                                                                                                         
> Out[48]: '5.8e-07'
> ```
> 
> What is it that you are not sure about? Is it common to use a comma with numbers in this form? Or is it that you think `latex()` should show this with `\cdot 10^{-7}`?

It is my understanding that the LatexPrinter's main use case is to beautifully typeset a given Sympy expression, so I'd definitely prefer an output with `\cdot 10^{-7}`. But I don't think it should actively hide Sympy's internal inconsistencies. At least not by default and not when the LaTeX output is used to typeset the responses in a Jupyter notebook. An additional option to automatically suppress these rounding issues (e.g. systematically sacrifice the last digit in order to round from 5.599999999999999 to 5.6) when calling latex() would be much appreciated, of course, but that would probably only be a workaround and out of scope for my request.

In conclusion, and given the current floating point idiosyncrasies, which are independent of this feature request, I'd expect the following:
```
>>> latex(5.8*10**(-7), decimal_separator='comma')
'5{,}8 \\cdot 10^{-7}'
>>> latex(5.6*10**(-7), decimal_separator='comma')
'5{,}599999999999999 \\cdot 10^{-7}'
```

Thank you for your detailed and informative follow-up!

**Comment by oscarbenjamin** [comment]
> I'd expect the following:
> 
> ```
> >>> latex(5.8*10**(-7), decimal_separator='comma')
> '5{,}8 \\cdot 10^{-7}'
> >>> latex(5.6*10**(-7), decimal_separator='comma')
> '5{,}599999999999999 \\cdot 10^{-7}'
> ```

Looking at I'm not sure about this. The reason `latex` is currently outputting these with the 1e2 notation is that they are Python floats rather than SymPy objects. I'm not sure if that is intentional. As you say this can also be used in Jupyter...

**Comment by oscargus** [comment]
After https://github.com/sympy/sympy/blob/d15b67d9a62e7528fb18fc29cc5817624a42ad45/sympy/printing/latex.py#L376
It should be possible to do something like
```
if self._settings['decimal_separator'] == 'comma':
    mant.replace('.', '{,}')
```
and before https://github.com/sympy/sympy/blob/d15b67d9a62e7528fb18fc29cc5817624a42ad45/sympy/printing/latex.py#L387
```
if self._settings['decimal_separator'] == 'comma':
    str_real.replace('.', '{,}')
```
(If may even be possible to just do the latter after   https://github.com/sympy/sympy/blob/d15b67d9a62e7528fb18fc29cc5817624a42ad45/sympy/printing/latex.py#L369 if we are sure that the only period is in the mantissa part, which I think it should be.)

In addition one will have to add `decimal_separator` to the default settings, to the function call, and the documentation. Possible also some check that `decimal_separator` takes on valid values.

**Comment by abhi-9810** [comment]
Hey, i am willing to work on this issue, i think it doesn't require much change. As mentioned by @oscargus we need to use "replace()". Please point out if i am missing anything as i am really new to this.

**Comment by oscargus** [comment]
@abhi-9810 please go ahead! There will be feedback on your pull request eventually.

**Comment by rimibis** [comment]
Hi! I'm not sure what the status of this issue is but I'm interested in trying to solve it. So I am currently trying to update the print_Float function by doing what @oscargus has suggested. However, I also wanted to implement the semicolons in the list request as well as the printing python decimal with commas(latex(0.98) prints 0{,}98) but I'm not sure where that would be handled. Would it fall under the EmptyPrinter? If so, I'm having a hard time figuring out where the EmptyPrinter code is. Any help would be appreciated!

Also, what does the S function in latex(**S**(0.987), decimal_separator='comma') stand for? Where is this function?

Thank you!

**Comment by rimibis** [comment]
Also @gschintgen for 

`>>> latex((1, 2.3, 4.5))
'\\left( 1; \\  2{,}3, \\  4{,}5\\right)'`

are you missing the decimal_separator option or is it supposed to be like that? Also, when I print, I print one back slash instead of two like you have in your output. I don't know if this is because of a setting that I accidentally turned on but I just wanted to see if anyone else was getting the same issue. Thanks!

**Comment by gschintgen** [comment]
> Also, what does the S function in latex(**S**(0.987), decimal_separator='comma') stand for? Where is this function?

Here it seems:
[https://github.com/sympy/sympy/blob/34e4ab88c866e15ccc6670bd8d5525e4102ed081/sympy/core/singleton.py#L123](
https://github.com/sympy/sympy/blob/34e4ab88c866e15ccc6670bd8d5525e4102ed081/sympy/core/singleton.py#L123)
It's a kind of shorthand for sympify. (as singleton object)

> 
> Also @gschintgen for
> 
> >>> latex((1, 2.3, 4.5)) '\\left( 1; \\ 2{,}3, \\ 4{,}5\\right)'
> 
> are you missing the decimal_separator option or is it supposed to be like that?

Yes, while editing the fictitious example I missed this line.

> 
> Also, when I print, I print one back slash instead of two like you have in your output. Even in the source code, I see the following:
> 
> return r"\left( %s\right)" % \ r", \ ".join([self._print(i) for i in expr])
> 
> which seems to indicate one back slash? I just wanted to see if this was an issue others were having?

That probably depends on the representation (raw, vs with escapes etc.). E.g.:
```
Python 3.6.7 (default, Oct 22 2018, 11:32:17)
[GCC 8.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> a = r'\hi'
>>> a
'\\hi'
```

So one backslash in the raw string is just fine, the printing of the second backslash is a necessary "artifact" of Python printing as regular string.

**Comment by rimibis** [comment]
Thanks! I was able to output what you wanted for the examples. I'm working on testing and should have a pull request up in a few hours. The samples you've given all work and I've also tried some other expressions(such as multplication and additions and such) and they all seem to work.

**Comment by smichr** [comment]
> This is compounded by a bug in SymPy's `Float.__str__`:

Is there are bug? These are two different numbers:
```console
>>> (5.6*10**-7).as_integer_ratio()
(2644525230407001, 4722366482869645213696)
>>> (5.6e-7).as_integer_ratio()
(5289050460814003, 9444732965739290427392)
```

**Comment by oscarbenjamin** [comment]
> Is there are bug?

The bug is that unequal numbers have the same str/repr. The two Python floats you've shown have different reprs:
```python
>>> 5.6*10**-7
5.599999999999999e-07
>>> 5.6e-7
5.6e-07
```
isympy's printing apparently uses the str for these rather than repr:
```julia
In [1]: 5.6*10**-7
Out[1]: 5.6e-07

In [2]: 5.6e-7
Out[2]: 5.6e-07

In [3]: 5.6*10**-7 == 5.6e-7
Out[3]: False
```

With SymPy's Floats the two unequal numbers can have the same str and repr:
```julia
In [4]: x = S(5.6e-7)

In [5]: y = S(5.6*10**-7)

In [6]: x
Out[6]: 5.60000000000000e-7

In [7]: y
Out[7]: 5.60000000000000e-7

In [8]: repr(x)
Out[8]: '5.60000000000000e-7'

In [9]: repr(y)
Out[9]: '5.60000000000000e-7'

In [10]: x == y
Out[10]: False

In [11]: repr(x) == repr(y)
Out[11]: True
```

**Comment by rimibis** [comment]
Is the float mismatch a separate issue? I haven't addressed it in my patch. 
Also @gschintgen did you want to implement this semicolon option in other places that use commas as well(such as f(x, y)). As of now, I've only implemented the semicolon for lists, sets, and tuples(just as you've listed in your examples).

**Comment by Sc0rpi0n101** [comment]
I think the feature is to display decimal numbers with comma separators instead of periods as some conventions might prefer using a comma for decimal numbers.

It was required in lists, sets, and tuples because you can't just leave it  like:
```python
[2,345 , 4,456 , 5,675]
```
It's very confusing. You have to do something like:
```python
[2,345 ; 4,456 ; 5,576]
```

But does the other places that us comma affect it? I don't think people will have any problem with things like f(x, y) when using decimal outputs with commas. Is there really a need to change these?

It might be better to find and change stuff that represents decimals/floats separated by commas

**Comment by oscargus** [comment]
I'd say implement it in the float printing and postpone possible related issues like list printing etc.

**Comment by rimibis** [comment]
I think that we should leave the f(x,y) as is since I think it's a common notation across situations. Please correct me if I'm wrong.

The pull request I've submitted contains the feature request changes for both floats and lists,sets, and tuples.

**Comment by gschintgen** [comment]
Well, by default a comma would be used in lists, sets, etc. but this comma is replaced by a semicolon if there's a possible confusion. E.g. in a 2-dimensional geometric context you could have in the same text the points A(x, y), B(2, -3) but C(1,2; 3,4).

But such a convention would be somewhat strange for a CAS. I guess most users expect a more deterministic output. That's why I tentatively proposed to generally replace the comma by a semicolon in lists, couples, sets. But @asmeurer  is of course right, that the same issue can arise in f(x,y).

I can't give a definite answer for this last case. It'd be logical to change it too, but in lists or sets as output, floats may appear a lot more frequently. Personally I'd be fine with @rimibis 's proposal and pull request. (Thanks!!)

Anyway, the most precise control over typesetting can only be achieved by manual editing of the LaTeX code. IMHO Sympy's output should adhere to reasonable defaults, even if they're sometimes "only" a compromise.

**Comment by rimibis** [comment]
Hello again! I was review some of the documentation on the PR I submitted for this issue. I just realised I haven't been printing the single quote around the entire expression:

my output: \left[ 1; \  2{,}3; \  4{,}5\right]
your output: '\left[ 1; \  2{,}3; \  4{,}5\right]'

For some reason, I missed this and was wondering if you wanted the quotes?

**Comment by oscargus** [comment]
The single ticks are to indicate that it is a string, so your output should be OK.

**Comment by rimibis** [comment]
Also, in the example 
`>>> latex((1, 2.3, 4.5), decimal_separator="comma") `
The output I am printing out has the decimal periods replaced be commas AND all the commas separating the elements in the tuple replaced by semicolons.
In the example give by gschintgen, they have:
`'\left( 1; \ 2{,}3, \ 4{,}5\\right)'`
I print:
\left( 1; \  2{,}3; \  4{,}5\right)

This might be a typo but in case it isn't I just wanted to point this out as we get closer to the final stages of the PR. Thanks!

**Comment by gschintgen** [comment]
Hi, I've set up sympy based on the current state of the pull request in a new LXD-container, together with LaTeX and pythontex. It works as intendend and wished for :-)

@rimibis : No change in single quotes or backslashes is wanted or asked for. The comma instead of semicolon that you spotted in the list example I gave is indeed an unfortunate typo. The requested change is *only* about the comma in floating point numbers and the list separator in lists, tuples, sets.
